### ISSUE TITLE
Entrypoint script for environments that need prior installation

### ISSUE THREAD

**Issue by garethr** [garethr/snyk-images#1]
Some environments are fine without any commands run, but others require some work before running `snyk test`. This should be simple enough to determine and encapsulate in a single entrypoint script.

Work without an entry script:

* [x] Node
* [x] Go modules
* [x] PHP
* [x] Ruby
* [x] Gradle
* [x] .NET
* [x] Scala/SBT

Need some entry script:

* [x] Python
* [x] Maven

**Comment by garethr** [comment]
PHP works without an entrypoint

```console
$ git clone git@github.com:gothinkster/laravel-realworld-example-app.gi
$ docker run --rm -it --env SNYK_TOKEN -v (pwd):/app snyk/snyk-composer test --file=composer.loc

Testing /app...

✗ Low severity vulnerability found in laravel/framework
  Description: Encryption Failure
  Info: https://snyk.io/vuln/SNYK-PHP-LARAVELFRAMEWORK-72114
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23
  Fixed in: 5.5.40, 5.6.15

✗ Medium severity vulnerability found in symfony/http-foundation
  Description: Denial of Service (DoS)
  Info: https://snyk.io/vuln/SNYK-PHP-SYMFONYHTTPFOUNDATION-72204
  Introduced through: barryvdh/laravel-cors@0.9.2, laravel/framework@5.4.23
  From: barryvdh/laravel-cors@0.9.2 > symfony/http-foundation@3.2.8
  From: laravel/framework@5.4.23 > symfony/http-foundation@3.2.8
  From: barryvdh/laravel-cors@0.9.2 > symfony/http-kernel@3.2.8 > symfony/http-foundation@3.2.8
  and 1 more...
  Fixed in: 2.7.48, 2.8.41, 3.3.17, 3.4.11, 4.0.11

✗ Medium severity vulnerability found in symfony/http-foundation
  Description: Access Restriction Bypass
  Info: https://snyk.io/vuln/SNYK-PHP-SYMFONYHTTPFOUNDATION-72253
  Introduced through: barryvdh/laravel-cors@0.9.2, laravel/framework@5.4.23
  From: barryvdh/laravel-cors@0.9.2 > symfony/http-foundation@3.2.8
  From: laravel/framework@5.4.23 > symfony/http-foundation@3.2.8
  From: barryvdh/laravel-cors@0.9.2 > symfony/http-kernel@3.2.8 > symfony/http-foundation@3.2.8
  and 1 more...
  Fixed in: 2.7.49, 2.8.44, 3.3.18, 3.4.14, 4.0.14, 4.1.3

✗ Medium severity vulnerability found in laravel/framework
  Description: Remote Code Execution (RCE)
  Info: https://snyk.io/vuln/SNYK-PHP-LARAVELFRAMEWORK-174581
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23
  Fixed in: 5.6.30, 5.5.41

✗ Medium severity vulnerability found in laravel/framework
  Description: Timing Attack
  Info: https://snyk.io/vuln/SNYK-PHP-LARAVELFRAMEWORK-72071
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23
  Fixed in: 5.5.10

✗ Medium severity vulnerability found in laravel/framework
  Description: Cookie serialization vulnerability
  Info: https://snyk.io/vuln/SNYK-PHP-LARAVELFRAMEWORK-72264
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23
  Fixed in: 5.6.30, 5.5.41

✗ Medium severity vulnerability found in erusev/parsedown
  Description: Cross-site Scripting (XSS)
  Info: https://snyk.io/vuln/SNYK-PHP-ERUSEVPARSEDOWN-72082
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23 > erusev/parsedown@1.6.2
  Fixed in: 1.7.0

✗ Medium severity vulnerability found in erusev/parsedown
  Description: Cross-site Scripting (XSS)
  Info: https://snyk.io/vuln/SNYK-PHP-ERUSEVPARSEDOWN-72138
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23 > erusev/parsedown@1.6.2
  Fixed in: 1.7.0

✗ High severity vulnerability found in laravel/framework
  Description: Information Exposure
  Info: https://snyk.io/vuln/SNYK-PHP-LARAVELFRAMEWORK-72070
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23
  Fixed in: 5.5.22

✗ High severity vulnerability found in erusev/parsedown
  Description: Arbitrary Code Injection
  Info: https://snyk.io/vuln/SNYK-PHP-ERUSEVPARSEDOWN-174118
  Introduced through: laravel/framework@5.4.23
  From: laravel/framework@5.4.23 > erusev/parsedown@1.6.2
  Fixed in: 1.7.2



Organization:      garethr
Package manager:   composer
Target file:       composer.lock
Open source:       no
Project path:      /app
Licenses:          enabled

Tested 46 dependencies for known issues, found 10 issues, 16 vulnerable paths.
```

**Comment by garethr** [comment]
Ruby projects work where a `Gemfile.lock` exists. Which is good enough to begin with, but might be nice in the entrypoint script.

```console
$ git clone git@github.com:puppetlabs/razor-server.git
$ docker run --rm -it --env SNYK_TOKEN -v (pwd):/app snyk/snyk-ruby

Testing /app...

✗ Medium severity issue found in gettext
  Description: LGPL-3.0 license
  Info: https://snyk.io/vuln/snyk:lic:rubygems:gettext:LGPL-3.0
  Introduced through: gettext-setup@0.30
  From: gettext-setup@0.30 > gettext@3.2.9



Organization:      garethr
Package manager:   rubygems
Target file:       Gemfile
Open source:       no
Project path:      /app
Licenses:          enabled

Tested 53 dependencies for known issues, found 1 issue, 1 vulnerable path.
```

**Comment by garethr** [comment]
Gradle works

```console
$ git clone git@github.com:gradle/gradle-build-scan-quickstart.git
$ docker run --rm -it --env SNYK_TOKEN -v (pwd):/app snyk/snyk-gradle
Testing /app...

✗ Medium severity issue found in junit:junit
  Description: EPL-1.0 license
  Info: https://snyk.io/vuln/snyk:lic:maven:junit:junit:EPL-1.0
  Introduced through: junit:junit@4.12
  From: junit:junit@4.12

✗ Medium severity vulnerability found in com.google.guava:guava
  Description: Deserialization of Untrusted Data
  Info: https://snyk.io/vuln/SNYK-JAVA-COMGOOGLEGUAVA-32236
  Introduced through: com.google.guava:guava@23.0
  From: com.google.guava:guava@23.0
  Remediation:
    Upgrade direct dependency com.google.guava:guava@23.0 to com.google.guava:guava@24.1.1-android (triggers upgrades to com.google.guava:guava@24.1.1-android)



Organization:      garethr
Package manager:   gradle
Target file:       build.gradle
Open source:       no
Project path:      /app
Licenses:          enabled

Tested 7 dependencies for known issues, found 2 issues, 2 vulnerable paths.
```

**Comment by garethr** [comment]
.NET works:

```console
$ dotnet new web
$ cd web
$ docker run --rm -it --env SNYK_TOKEN -v (pwd):/app snyk/snyk-mcr.microsoft.com/dotnet/core/sdk
...

✗ High severity vulnerability found in Microsoft.NETCore.App
  Description: Domain-Spoofing
  Info: https://snyk.io/vuln/SNYK-DOTNET-MICROSOFTNETCOREAPP-173690
  Introduced through: Microsoft.NETCore.App@2.2.0
  From: Microsoft.NETCore.App@2.2.0
  Fixed in: 2.1.8, 2.2.2

✗ High severity vulnerability found in Microsoft.NETCore.App
  Description: Authentication Bypass by Spoofing
  Info: https://snyk.io/vuln/SNYK-DOTNET-MICROSOFTNETCOREAPP-173754
  Introduced through: Microsoft.NETCore.App@2.2.0
  From: Microsoft.NETCore.App@2.2.0
  Fixed in: 2.1.8, 2.2.2

✗ High severity vulnerability found in Microsoft.NETCore.App
  Description: Denial of Service (DoS)
  Info: https://snyk.io/vuln/SNYK-DOTNET-MICROSOFTNETCOREAPP-174710
  Introduced through: Microsoft.NETCore.App@2.2.0
  From: Microsoft.NETCore.App@2.2.0
  Fixed in: 2.2.5

✗ High severity vulnerability found in Microsoft.AspNetCore.WebSockets
  Description: Denial of Service (DoS)
  Info: https://snyk.io/vuln/SNYK-DOTNET-MICROSOFTASPNETCOREWEBSOCKETS-72898
  Introduced through: Microsoft.AspNetCore.App@2.2.0
  From: Microsoft.AspNetCore.App@2.2.0 > Microsoft.AspNetCore.WebSockets@2.2.0
  From: Microsoft.AspNetCore.App@2.2.0 > Microsoft.AspNetCore.Http.Connections@1.1.0 > Microsoft.AspNetCore.WebSockets@2.2.0
  From: Microsoft.AspNetCore.App@2.2.0 > Microsoft.AspNetCore.SpaServices.Extensions@2.2.0 > Microsoft.AspNetCore.WebSockets@2.2.0
  and 1 more...
  Fixed in: 2.1.7, 2.2.1

✗ High severity vulnerability found in Microsoft.AspNetCore
  Description: Denial of Service (DOS)
  Info: https://snyk.io/vuln/SNYK-DOTNET-MICROSOFTASPNETCORE-174184
  Introduced through: Microsoft.AspNetCore.App@2.2.0
  From: Microsoft.AspNetCore.App@2.2.0 > Microsoft.AspNetCore@2.2.0

Organization:      garethr
Package manager:   nuget
Target file:       obj/project.assets.json
Open source:       no
Project path:      /app
Licenses:          enabled

Tested 284 dependencies for known issues, found 18 issues, 584 vulnerable paths.
```

**Comment by garethr** [comment]
Scala works, however their are no official or well supported images. This appears to be the latest updated https://hub.docker.com/r/hseeberger/scala-sbt/tags, but only has tags for versions.

Note that the project must have the `sbt-dependency-graph` plugin in the plugins file too.

```console
$ git clone git@github.com:lampepfl/dotty-example-project.git
$ echo `addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.10.0-RC1")` >> project/plugins.sbt
$ docker run --rm -it --env SNYK_TOKEN -v (pwd):/app snyk/snyk-hseeberger/scala-sbt:8u212_1.2.8_2.13.0

Testing /app...

Organization:      garethr
Package manager:   sbt
Target file:       build.sbt
Open source:       no
Project path:      /app
Licenses:          enabled

✓ Tested 1 dependencies for known issues, no vulnerable paths found.

Next steps:
- Run `snyk monitor` to be notified about new related vulnerabilities.
- Run `snyk test` as part of your CI/test
```

**Comment by garethr** [comment]
Python working with entrypoint script, for Pipenv and requirements.txt.

```console
$ docker run --rm -it --env SNYK_TOKEN -v (pwd):/app snyk/snyk-python
Collecting pipenv
  Downloading https://files.pythonhosted.org/packages/13/b4/3ffa55f77161cff9a5220f162670f7c5eb00df52e00939e203f601b0f579/pipenv-2018.11.26-py3-none-any.whl (5.2MB)
     |████████████████████████████████| 5.2MB 1.3MB/s
Requirement already satisfied: setuptools>=36.2.1 in /usr/local/lib/python3.7/site-packages (from pipenv) (41.2.0)
Collecting certifi (from pipenv)
  Downloading https://files.pythonhosted.org/packages/69/1b/b853c7a9d4f6a6d00749e94eb6f3a041e342a885b87340b79c1ef73e3a78/certifi-2019.6.16-py2.py3-none-any.whl (157kB)
     |████████████████████████████████| 163kB 3.2MB/s
Collecting virtualenv-clone>=0.2.5 (from pipenv)
  Downloading https://files.pythonhosted.org/packages/ba/f8/50c2b7dbc99e05fce5e5b9d9a31f37c988c99acd4e8dedd720b7b8d4011d/virtualenv_clone-0.5.3-py2.py3-none-any.whl
Collecting virtualenv (from pipenv)
  Downloading https://files.pythonhosted.org/packages/8b/12/8d4f45b8962b03ac9efefe5ed5053f6b29334d83e438b4fe379d21c0cb8e/virtualenv-16.7.5-py2.py3-none-any.whl (3.3MB)
     |████████████████████████████████| 3.3MB 11.7MB/s
Requirement already satisfied: pip>=9.0.1 in /usr/local/lib/python3.7/site-packages (from pipenv) (19.2.3)
Installing collected packages: certifi, virtualenv-clone, virtualenv, pipenv
Successfully installed certifi-2019.6.16 pipenv-2018.11.26 virtualenv-16.7.5 virtualenv-clone-0.5.3
Creating a virtualenv for this project…
Pipfile: /app/Pipfile
Using /usr/local/bin/python (3.7.4) to create virtualenv…
⠏ Creating virtual environment...Already using interpreter /usr/local/bin/python
Using base prefix '/usr/local'
New python executable in /root/.local/share/virtualenvs/app-4PlAip0Q/bin/python
Installing setuptools, pip, wheel...
done.

✔ Successfully created virtual environment!
Virtualenv location: /root/.local/share/virtualenvs/app-4PlAip0Q
Installing dependencies from Pipfile.lock (04695c)…
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 7/7 — 00:00:01
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
All dependencies are now up-to-date!

Testing /app...

✗ High severity vulnerability found in flask
  Description: Improper Input Validation
  Info: https://snyk.io/vuln/SNYK-PYTHON-FLASK-42185
  Introduced through: flask@0.12
  From: flask@0.12
  Fixed in: 0.12.3

✗ High severity vulnerability found in flask
  Description: Denial of Service (DOS)
  Info: https://snyk.io/vuln/SNYK-PYTHON-FLASK-451637
  Introduced through: flask@0.12
  From: flask@0.12
  Fixed in: 1.0



Organization:      garethr
Package manager:   pip
Target file:       Pipfile
Open source:       no
Project path:      /app
Licenses:          enabled

Tested 7 dependencies for known issues, found 2 issues, 2 vulnerable paths.
```

**Comment by garethr** [comment]
Working from https://github.com/garethr/snyk-images/commit/383a441bb69d2169efc57ef04cab62e2a5d57a4d
